# Tarea 3: Sequence Labelling
**Procesamiento de Lenguaje Natural (CC6205-1 - Otoño 2024)**

## Tarjeta de identificación

**Nombres:**

```- Ignacio Albornoz```

```- Eduardo Silva```

**Fecha límite de entrega 📆:** 06/06.

**Tiempo estimado de dedicación:** 4 horas


## Instrucciones
Bienvenid@s a la tercera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las últimas semanas de clases posteriores a la tarea 2, enfocado principalmente en **Word Embeddings**, **Sequence Labeling - HMM**, **Convolutional Neural Networks** y **Recurrent Neural Networks**. Si aún no has visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP.

* La tarea es en **grupo** (maximo hasta 3 personas).
* La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
* El formato de entrega es este mismo Jupyter Notebook.
* Al momento de la revisión su código será ejecutado. Por favor verifiquen que su entrega no tenga errores de compilación.
* Completar la tarjeta de identificación. Sin ella no podrá tener nota.

## Material de referencia

Diapositivas del curso 📄
    
- [Word Embeddings](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-wordvectors.pdf)
- [Sequence Labeling - HMM](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-HMM.pdf)
- [Convolutional Neural Networks](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-CNN.pdf)
- [Recurrent Neural Networks](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-RNN.pdf)

Videos del curso 📺

- Word Embeddings: [Parte 1](https://www.youtube.com/watch?v=wtwUsJMC9CA), [Parte 2](https://www.youtube.com/watch?v=XDxzQ7JU95U), [Parte 3](https://www.youtube.com/watch?v=Ikyc3DRVodk)

- Sequence Labeling - HMM: [Parte 1](https://www.youtube.com/watch?v=-ngfOZz8yK0), [Parte 2](https://www.youtube.com/watch?v=Tjgb-yQOg54), [Parte 3](https://www.youtube.com/watch?v=aaa5Qoi8Vco), [Parte 4](https://www.youtube.com/watch?v=4pKWIDkF_6Y)

- [Convolutional Neural Networks](https://www.youtube.com/watch?v=lLZW5Fn40r8)

- Recurrent Neural Networks: [Parte 1](https://www.youtube.com/watch?v=BmhjUkzz3nk), [Parte 2](https://www.youtube.com/watch?v=z43YFR1iIvk), [Parte 3](https://youtu.be/7L5JxQdwNJk)

## Objetivo

El objetivo de esta tarea es resolver una de las tareas más importantes en el área del procesamiento de lenguage natural, relacionada con la extracción de información: [Named Entity Recognition (NER)](http://www.cs.columbia.edu/~mcollins/cs4705-spring2019/slides/tagging.pdf).

En particular, deberán crear distintos modelos que apunten a resolver la tarea de NER en Español. Para esto, les entregaremos un dataset real perteneciente a la lista de espera NO GES en Chile. Es importante destacar que existe una falta de trabajos realizados en el área de NER en Español y aún más en el contexto clínico, por ende puede ser considerado como una tarea bien desafiante y quizás les interesa trabajar en el área más adelante en sus carreras.

En este notebook les entregaremos un baseline como referencia de los resultados que esperamos puedan obtener. Recuerden que el no superar a los baselines en alguna de las tres métricas conlleva un descuento de 0.5 puntos hasta 1.5 puntos.

Como hemos estado viendo redes neuronales tanto en cátedras, tareas y auxiliares (o próximamente lo harán), esperamos que (por lo menos) utilicen Redes Neuronales Recurrentes (RNN) para resolverla.

Nuevamente, hay total libertad para utilizar el software y los modelos que deseen, siempre y cuando estos no traigan los modelos ya implementados (de todas maneras, como es un corpus nuevo, es difícil que haya algún modelo ya implementado con éstas entidades).

## Explicación de NER

En esta tarea van a resolver **NER**, comúnmente abordada como un problema de Sequence Labeling.

**¿Qué es Sequence Labeling?**

En breves palabras, dada una secuencia de tokens (oración) sequence labeling tiene por objetivo asignar una etiqueta a cada token de dicha secuencia. En pocas palabras, dada una lista de tokens esperamos encontrar la mejor secuencia de etiquetas asociadas a esa lista. Ahora veamos de qué se trata este problema.

**Named Entity Recognition (NER)**

NER es un ejemplo de un problema de Sequence Labeling. Pero antes de definir formalmente esta tarea, es necesario definir algunos conceptos claves para poder entenderla de la mejor manera:

- *Token*: Un token es una secuencia de caracteres, puede ser una palabra, un número o un símbolo.

- *Entidad*: No es más que un trozo de texto (uno o más tokens) asociado a una categoría predefinida. Originalmente se solían utilizar categorías como nombres de personas, organizaciones, ubicaciones, pero actualmente se ha extendido a diferentes dominios.

- *Límites de una entidad*: Son los índices de los tokens de inicio y fín dentro de una entidad.

- *Tipo de entidad*: Es la categoría predefinida asociada a la entidad.

Dicho esto, definimos formalmente una entidad como una tupla: $(s, e, t)$, donde $s, $e son los límites de la entidad (índices de los tokens de inicio y fin, respectivamente) y $t corresponde al tipo de entidad o categoría. Ya veremos más ejemplos luego de describir el Dataset.

**Corpus de la Lista de espera**

Trabajaran con un conjunto de datos reales correspondiente a interconsultas de la lista de espera NO GES en Chile. Si quieren saber más sobre cómo fueron generados los datos pueden revisar el paper publicado hace unos meses atrás en el workshop de EMNLP, una de las conferencias más importantes de NLP: [https://www.aclweb.org/anthology/2020.clinicalnlp-1.32/](https://www.aclweb.org/anthology/2020.clinicalnlp-1.32/).

Este corpus Chileno está constituido originalmente por 7 tipos de entidades pero por simplicidad en esta tarea trabajarán con las siguientes:

- **Disease**
- **Body_Part**
- **Medication**
- **Procedures**
- **Family_Member**

Si quieren obtener más información sobre estas entidades pueden consultar la [guía de anotación](https://plncmm.github.io/annodoc/). Además, mencionar que este corpus está restringido bajo una licencia que permite solamente su uso académico, así que no puede ser compartido más allá de este curso o sin permisos por parte de los autores en caso que quieran utilizarlo fuera. Si este último es el caso entonces pueden escribir directamente al correo: pln@cmm.uchile.cl. Al aceptar los términos y condiciones de la tarea están de acuerdo con los puntos descritos anteriormente.


**Formato ConLL**

Los archivos que serán entregados a ustedes vienen en un formato estándar utilizado en NER, llamado ConLL. No es más que un archivo de texto, que cumple las siguientes propiedades.

- Un salto de linea corresponde a la separación entre oraciones. Esto es importante ya que al entrenar una red neuronal ustedes pasaran una lista de oraciones como input, más conocidos como batches.

- La primera columna del archivo contiene todos los tokens de la partición.

- La segunda columna del archivo contiene el tipo de entidad asociado al token de la primera columna.

- Los tipos de entidades siguen un formato clásico en NER denominado *IOB2*. Si un tipo de entidad comienza con el prefijo "B-" (Beginning) significa que es el token de inicio de una entidad, si comienza con "I-" (Inside) es un token distinto al de inicio y si un token está asociado a la categoría O (Outside) significa que no pertenece a ninguna entidad.

Aquí va un ejemplo:

```
PACIENTE O
PRESENTA O
FRACTURA B-Disease
CORONARIA I-Disease
COMPLICADA I-Disease
EN O
PIE B-Body_Part
IZQUIERDO I-Body_Part
. O
SE O
REALIZA O
INSTRUMENTACION B-Procedure
INTRACONDUCTO I-Procedure
. O
```

Según nuestra definición tenemos las siguientes tres entidades (enumerando desde 0):

- $(2, 4, Disease)$
- $(6, 7, Body Part)$
- $(11, 12, Procedure)$

Repasen un par de veces todos estos conceptos antes de pasar a la siguiente sección del notebook.
Es importante entender bien este formato ya que al medir el rendimiento de sus modelos, consideraremos una **métrica estricta**. Esta métrica se llama así ya que considera correcta una predicción de su modelo, sólo si al compararlo con las entidades reales **coinciden tanto los límites de la entidad como el tipo.**

Para ejemplificar, tomando el caso anterior, si el modelo es capaz de encontrar la siguiente entidad: $(2, 3, Disease)$, entonces se considera incorrecto ya que pudo predecir dos de los tres tokens de dicha enfermedad. Es decir, buscamos una métrica que sea alta a nivel de entidad y no a nivel de token.

Antes de pasar a explicar las reglas, se recomienda visitar los siguientes links para entender bien el baseline de la tarea:

-  [Recurrent Neural Networks](slides/NLP-RNN.pdf): [Parte 1](https://youtu.be/BmhjUkzz3nk), [Parte 2](https://youtu.be/z43YFR1iIvk), [Parte 3](https://youtu.be/7L5JxQdwNJk)


Recuerden que todo el material se encuentra disponible en el [github del curso](https://github.com/dccuchile/CC6205).

## Baseline

En este punto esperamos que tengan conocimiento sobre redes neuronales y en particular redes neuronales recurrentes (RNN), si no siempre pueden escribirnos por el canal de Discord para aclarar dudas. La RNN del baseline adjunto a este notebook está programado en la librería [`pytorch`](https://pytorch.org/) pero ustedes pueden utilizar keras, tensorflow si así lo desean. El código contiene lo siguiente:

- La carga de los datasets, creación de batches de texto y padding (esto es importante ya que si utilizan redes neuronales tienen que tener el mismo largo los inputs).

- La implementación básica de una red `LSTM` simple de solo un nivel y sin bi-direccionalidad.

- La construcción del formato del output requerido para que lo puedan probar en la tarea en codalab.

### **Sugerencias**
Se espera que ustedes puedan experimentar con el baseline utilizando (pero no limitándose) estas sugerencias:

*   Probar la técnica de early stopping.
*   Variar la cantidad de parámetros de la capa de embeddings.
*   Variar la cantidad de capas RNN.
*   Variar la cantidad de parámetros de las capas de RNN.
*   Inicializar la capa de embeddings con modelos pre-entrenados. (word2vec, glove, conceptnet, etc...). [Embeddings en español aquí](https://github.com/dccuchile/spanish-word-embeddings). También aquí pueden encontrar unos embeddings clínicos en Español: [https://zenodo.org/record/3924799](https://zenodo.org/record/3924799)
*   Variar la cantidad de épocas de entrenamiento.
*   Variar el optimizador, learning rate, batch size, etc.
*   Probar bi-direccionalidad.
*   Incluir dropout.
*   Probar modelos de tipo GRU.
*   Probar usando capas de atención.
*   Probar Embedding Contextuales (les puede ser de utilidad [flair](https://github.com/flairNLP/flair))
*   Probar modelos de transformers en español usando [Huggingface](https://github.com/huggingface/transformers) o el framework Flair.

## Experimento base

El código que les entregaremos servirá de baseline para luego implementar mejores modelos.
En general, el código asociado a la carga de los datos, las funciones de entrenamiento, de evaluación y la predicción de los datos de la tarea no deberían cambiar.
Solo deben preocuparse de cambiar la arquitectura del modelo, sus hiperparámetros y reportar, lo cual lo pueden hacer en las subsecciones *modelos*.

###  **Carga de datos y Preprocesamiento**

Para cargar los datos y preprocesarlos usaremos la librería [`torchtext`](https://github.com/pytorch/text). Tener cuidado ya que hace algunos meses esta librería tuvo cambios radicales, quedando las funcionalidades pasadas depreciadas de la librería ```legacy```. Esto ya que si quieren usar más funciones de la librería entonces vean los cambios en la documentación debe usar la versión antigua con python 3.8

El proceso será el siguiente:

1. Descargar los datos desde github y examinarlos.
2. Cargar los datasets con la clase ```TaggingDataset``` de más abajo.
3. Crear el vocabulario.

In [1]:
!pip install -U torchtext

  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)
  Using cached nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl (176.2 MB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (99 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.3/21

In [2]:
import torch
import torchtext
from torchtext import data, datasets

# Garantizar reproducibilidad de los experimentos
SEED = 1234
torch.manual_seed(SEED)
torch.backends.cudnn.deterministic = True

/usr/local/lib/python3.10/dist-packages/torchtext/data/__init__.py:4: UserWarning: 
/!\ IMPORTANT WARNING ABOUT TORCHTEXT STATUS /!\ 
Torchtext is deprecated and the last released version will be 0.18 (this one). You can silence this warning by calling the following at the beginnign of your scripts: `import torchtext; torchtext.disable_torchtext_deprecation_warning()`
  warnings.warn(torchtext._TORCHTEXT_DEPRECATION_MSG)
/usr/local/lib/python3.10/dist-packages/torchtext/datasets/__init__.py:4: UserWarning: 
/!\ IMPORTANT WARNING ABOUT TORCHTEXT STATUS /!\ 
Torchtext is deprecated and the last released version will be 0.18 (this one). You can silence this warning by calling the following at the beginnign of your scripts: `import torchtext; torchtext.disable_torchtext_deprecation_warning()`
  warnings.warn(torchtext._TORCHTEXT_DEPRECATION_MSG)


#### **Obtener datos**

Descargamos los datos de entrenamiento, validación y prueba en nuestro directorio de trabajo

In [3]:
%%capture

!wget https://github.com/dccuchile/CC6205/releases/download/v1.0/train.txt -nc
!wget https://github.com/dccuchile/CC6205/releases/download/v1.0/dev.txt -nc

In [4]:
# NUEVO DATALOADER Y OTRAS COSAS NECESARIAS
from collections import Counter, OrderedDict

import torch

from torch.utils.data import DataLoader, Dataset
from torch.nn.utils.rnn import pad_sequence
from torchtext.vocab import vocab

class TaggingDataset(Dataset):
    def __init__(self, paths, lower=False, separator=" ", encoding="utf-8"):

        data = []
        for path in paths:
          with open(path, 'r', encoding=encoding) as file:
            text, tag = [], []
            for line in file:
                line = line.strip()
                if line == "":
                    data.append(dict({'text':text, 'nertags':tag}))
                    text, tag = [], []
                else:
                    line_content = line.split(separator) # .rstrip('\n')
                    if lower:
                      text.append(line_content[0].lower())
                    else:
                      text.append(line_content[0])
                    tag.append(line_content[1])
          data.append(dict({'text':text, 'nertags':tag}))

        self.data = data

    def __len__(self):
        return len(self.data)

    def __getitem__(self, index):
        item = self.data[index]
        text = item["text"]
        nertags = item["nertags"]
        return nertags, text

def fit_vocab(data_iter):

  def update_counter(counter_obj):
    sorted_by_freq_tuples = sorted(counter_obj.items(),
                                  key=lambda x: x[1],
                                  reverse=True)
    ordered_dict = OrderedDict(sorted_by_freq_tuples)
    return ordered_dict

  counter_1 = Counter()
  counter_2 = Counter()
  for _nertags, _text in data_iter:
    counter_1.update(_text)
    counter_2.update(_nertags)

  od1 = update_counter(counter_1)
  od2 = update_counter(counter_2)

  v1 = vocab(od1, specials=['<PAD>', '<unk>'])
  v1.set_default_index(v1["<unk>"])
  v2 = vocab(od2, specials=['<PAD>'])

  text_pipeline = lambda x: v1(x)
  nertags_pipeline = lambda x: v2(x)

  return text_pipeline, nertags_pipeline, v1, v2

def collate_batch(batch, nertags_pipeline, text_pipeline, device):
  nertags_list, text_list = [], []
  for _nertags, _text in batch:
    processed_nertags = torch.tensor(nertags_pipeline(_nertags),
                                     dtype=torch.int64)
    nertags_list.append(processed_nertags)
    processed_text = torch.tensor(text_pipeline(_text),
                                  dtype=torch.int64)
    text_list.append(processed_text)
  nertags_list = pad_sequence(nertags_list, batch_first=True).T
  text_list = pad_sequence(text_list, batch_first=True).T
  return nertags_list.to(device), text_list.to(device)

/usr/local/lib/python3.10/dist-packages/torchtext/vocab/__init__.py:4: UserWarning: 
/!\ IMPORTANT WARNING ABOUT TORCHTEXT STATUS /!\ 
Torchtext is deprecated and the last released version will be 0.18 (this one). You can silence this warning by calling the following at the beginnign of your scripts: `import torchtext; torchtext.disable_torchtext_deprecation_warning()`
  warnings.warn(torchtext._TORCHTEXT_DEPRECATION_MSG)
/usr/local/lib/python3.10/dist-packages/torchtext/utils.py:4: UserWarning: 
/!\ IMPORTANT WARNING ABOUT TORCHTEXT STATUS /!\ 
Torchtext is deprecated and the last released version will be 0.18 (this one). You can silence this warning by calling the following at the beginnign of your scripts: `import torchtext; torchtext.disable_torchtext_deprecation_warning()`
  warnings.warn(torchtext._TORCHTEXT_DEPRECATION_MSG)


In [5]:
data_iter = TaggingDataset(["train.txt", "dev.txt"])

data_length = len(data_iter)

train_length = int(data_length * 0.8)
dev_length = int(data_length * 0.1)
test_length = data_length - train_length - dev_length

train_iter, dev_iter, test_iter = torch.utils.data.random_split(
    data_iter,
     (train_length, dev_length, test_length),
    torch.Generator().manual_seed(42))

text_pipeline, nertags_pipeline, TEXT, NER_TAGS = fit_vocab(train_iter)

In [6]:
len(train_iter), len(dev_iter), len(test_iter)

(7132, 891, 893)

In [7]:
# seteamos algunos valores de interes
UNK_IDX = TEXT.vocab.get_stoi()['<unk>']
PAD_IDX = TEXT.vocab.get_stoi()['<PAD>']

PAD_TAG_IDX = NER_TAGS.get_stoi()['<PAD>']
O_TAG_IDX = NER_TAGS.vocab.get_stoi()['O']

In [8]:
BATCH_SIZE = 22

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print('Using', device)

fix_collocate_batch = lambda x: collate_batch(x, nertags_pipeline, text_pipeline, device)

dataloader_train = DataLoader(
    train_iter, batch_size=BATCH_SIZE, shuffle=False, collate_fn=fix_collocate_batch
)
dataloader_dev = DataLoader(
    dev_iter, batch_size=BATCH_SIZE, shuffle=False, collate_fn=fix_collocate_batch
)
dataloader_test = DataLoader(
    test_iter, batch_size=BATCH_SIZE, shuffle=False, collate_fn=fix_collocate_batch
)

Using cuda


In [9]:
example = next(iter(dataloader_train))
ner_tags_example = example[0]
text_example = example[1]

# revisamos el primer ejemplo
[NER_TAGS.vocab.get_itos()[j] for j in ner_tags_example[:, 1]], [TEXT.vocab.get_itos()[j] for j in text_example[:, 1]]

(['O',
  'O',
  'O',
  'O',
  'B-Disease',
  'I-Disease',
  'I-Disease',
  'I-Disease',
  'O',
  'O',
  'O',
  'O',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>'],
 ['SOLICITO',
  'EVALUACION',
  'Y',
  'TTO',
  'SUBLUXACION',
  'ATM',
  'IZQ',
  '.',
  ',',
  'DOLOR',
  'APERTURA',
  '.',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>'])

In [10]:
!pip install seqeval

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 494.8 kB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for seqeval: filename=seqeval-1.2.2-py3-none-any.whl size=16161 sha256=6f6efe9ffdd197408e5c9b9f0b540680e91b54657907a80f635185a226891c12
  Stored in directory: /root/.cache/pip/wheels/1a/67/4a/ad4082dd7dfc30f2abfe4d80a2ed5926a506eb8a972b4767fa
Successfully built seqeval


In [11]:
# Definimos las métricas

from seqeval.metrics import f1_score, precision_score, recall_score

def calculate_metrics(preds, y_true, pad_idx=PAD_TAG_IDX, o_idx=O_TAG_IDX):
    """
    Calcula precision, recall y f1 de cada batch.
    """

    # Obtener el indice de la clase con probabilidad mayor. (clases)
    y_pred = preds.argmax(dim=1, keepdim=True)

    # filtramos <pad> para calcular los scores.
    mask = [(y_true != pad_idx)]
    y_pred = y_pred[mask]
    y_true = y_true[mask]

    # traemos a la cpu
    y_pred = y_pred.view(-1).to('cpu').numpy()
    y_true = y_true.to('cpu').numpy()
    y_pred = [[NER_TAGS.vocab.get_itos()[v] for v in y_pred]]
    y_true = [[NER_TAGS.vocab.get_itos()[v] for v in y_true]]

    # calcular scores
    f1 = f1_score(y_true, y_pred, mode='strict')
    precision = precision_score(y_true, y_pred, mode='strict')
    recall = recall_score(y_true, y_pred, mode='strict')

    return precision, recall, f1

### **Modelo Baseline**

Teniendo ya cargado los datos, toca definir nuestro modelo. Este baseline tendrá una capa de embedding, unas cuantas LSTM y una capa de salida y usará dropout en el entrenamiento.

Este constará de los siguientes pasos:

1. Definir la clase que contendrá la red.
2. Definir los hiperparámetros e inicializar la red.
3. Definir el número de épocas de entrenamiento
4. Definir la función de loss.



Recomendamos que para experimentar, encapsules los modelos en una sola variable y luego la fijes en model para entrenarla

In [12]:
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim


# Definir la red
class NER_RNN(nn.Module):
    def __init__(self,
                 input_dim,
                 embedding_dim,
                 hidden_dim,
                 output_dim,
                 n_layers,
                 bidirectional,
                 dropout,
                 pad_idx):

        super().__init__()

        # Capa de embedding
        self.embedding = nn.Embedding(input_dim,
                                      embedding_dim,
                                      padding_idx=pad_idx,
                                      )

        # Capa LSTM
        self.lstm = nn.LSTM(embedding_dim,
                           hidden_dim,
                           num_layers=n_layers,
                           bidirectional=bidirectional,
                           dropout = dropout if n_layers > 1 else 0)

        # Capa de salida
        self.fc = nn.Linear(hidden_dim * 2 if bidirectional else hidden_dim,
                            output_dim)

        # Dropout
        self.dropout = nn.Dropout(dropout)

    def forward(self, text):

        #text = [sent len, batch size]

        # Convertir lo enviado a embedding
        # embedded = self.dropout(self.embedding(text))
        embedded = self.embedding(text)

        outputs, (hidden, cell) = self.lstm(embedded)
        #embedded = [sent len, batch size, emb dim]

        # Pasar los embeddings por la rnn (LSTM)

        #output = [sent len, batch size, hid dim * n directions]
        #hidden/cell = [n layers * n directions, batch size, hid dim]

        # Predecir usando la capa de salida.
        #predictions = self.fc(self.dropout(outputs))
        predictions = self.fc(outputs)
        #predictions = [sent len, batch size, output dim]

        return predictions

Hiperparámetros de la red

In [13]:
# tamaño del vocabulario. recuerden que la entrada son vectores bag of word(one-hot).
INPUT_DIM = len(TEXT.vocab)
EMBEDDING_DIM = 300  # dimensión de los embeddings.
HIDDEN_DIM = 256  # dimensión de la capas LSTM
OUTPUT_DIM = len(NER_TAGS.vocab)  # número de clases

N_LAYERS = 3  # número de capas.
DROPOUT = 0.6
BIDIRECTIONAL = False

# Creamos nuestro modelo.
baseline_model = NER_RNN(INPUT_DIM, EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM,
                         N_LAYERS, BIDIRECTIONAL, DROPOUT, PAD_IDX)

baseline_model_name = 'baseline'  # nombre que tendrá el modelo guardado...

baseline_n_epochs = 10

Definimos la función de loss

In [14]:
# Loss: Cross Entropy
TAG_PAD_IDX = NER_TAGS.vocab.get_stoi()['<PAD>']
baseline_criterion = nn.CrossEntropyLoss(ignore_index = TAG_PAD_IDX)

### **Entrenamos y evaluamos**


**Importante** : Fijen el modelo, el número de épocas de entrenamiento, la loss y el optimizador que usarán para entrenar y evaluar en las siguientes variables!!!

In [15]:
model = baseline_model
model_name = baseline_model_name
criterion = baseline_criterion
n_epochs = baseline_n_epochs

#### **Inicializamos la red**

Iniciamos los pesos de la red de forma aleatoria (Usando una distribución normal).

In [16]:
def init_weights(m):
    # Inicializamos los pesos como aleatorios
    for name, param in m.named_parameters():
        nn.init.normal_(param.data, mean=0, std=0.1)

    # Seteamos como 0 los embeddings de UNK y PAD.
    model.embedding.weight.data[UNK_IDX] = torch.zeros(EMBEDDING_DIM)
    model.embedding.weight.data[PAD_IDX] = torch.zeros(EMBEDDING_DIM)

model.apply(init_weights)

NER_RNN(
  (embedding): Embedding(16496, 300, padding_idx=0)
  (lstm): LSTM(300, 256, num_layers=3, dropout=0.6)
  (fc): Linear(in_features=256, out_features=12, bias=True)
  (dropout): Dropout(p=0.6, inplace=False)
)

In [17]:
def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f'El modelo actual tiene {count_parameters(model):,} parámetros entrenables.')

El modelo actual tiene 6,575,948 parámetros entrenables.


Notar que definimos los embeddings que representan a \<unk\> y \<pad\>  como [0, 0, ..., 0]

Definimos el optimizador

In [18]:
# Optimizador

optimizer = optim.Adam(model.parameters()) # SGD, AdamW

Enviamos el modelo a cuda

In [19]:
# Enviamos el modelo y la loss a cuda (en el caso en que esté disponible)
model = model.to(device)
criterion = criterion.to(device)

#### **Definimos el entrenamiento de la red**

Algunos conceptos previos:

- `epoch` : una pasada de entrenamiento completa de una dataset.
- `batch`: una fracción de la época. Se utilizan para entrenar mas rápidamente la red. (mas eficiente pasar n datos que uno en cada ejecución del backpropagation)

Esta función está encargada de entrenar la red en una época. Para esto, por cada batch de la época actual, predice los tags del texto, calcula su loss y luego hace backpropagation para actualizar los pesos de la red.

Observación: En algunos comentarios aparecerá el tamaño de los tensores entre corchetes

In [20]:
def train(model, iterator, optimizer, criterion):

    epoch_loss = 0
    epoch_precision = 0
    epoch_recall = 0
    epoch_f1 = 0

    model.train()

    # Por cada batch del iterador de la época:
    for tags, text in iterator:
        # Reiniciamos los gradientes calculados en la iteración anterior
        optimizer.zero_grad()

        #text = [sent len, batch size]

        # Predecimos los tags del texto del batch.
        predictions = model(text.to(device))

        #predictions = [sent len, batch size, output dim]
        #tags = [sent len, batch size]

        # Reordenamos los datos para calcular la loss
        predictions = predictions.view(-1, predictions.shape[-1])
        #ipdb.set_trace()
        tags = torch.reshape(tags, (-1,)).to(device)

        #predictions = [sent len * batch size, output dim]

        # Calculamos el Cross Entropy de las predicciones con respecto a las etiquetas reales
        loss = criterion(predictions, tags)

        # Calculamos el accuracy
        precision, recall, f1 = calculate_metrics(predictions, tags)

        # Calculamos los gradientes
        loss.backward()

        # Actualizamos los parámetros de la red
        optimizer.step()

        # Actualizamos el loss y las métricas
        epoch_loss += loss.item()
        epoch_precision += precision
        epoch_recall += recall
        epoch_f1 += f1

    return ( epoch_loss / len(iterator), epoch_precision / len(iterator),
              epoch_recall / len(iterator), epoch_f1 / len(iterator) )

#### **Definimos la función de evaluación**

Evalua el rendimiento actual de la red usando los datos de validación.

Por cada batch de estos datos, calcula y reporta el loss y las métricas asociadas al conjunto de validación.
Ya que las métricas son calculadas por cada batch, estas son retornadas promediadas por el número de batches entregados. (ver linea del return)

In [21]:
def evaluate(model, iterator, criterion):

    epoch_loss = 0
    epoch_precision = 0
    epoch_recall = 0
    epoch_f1 = 0

    model.eval()

    # Indicamos que ahora no guardaremos los gradientes
    with torch.no_grad():
        # Por cada batch
        for tags, text in iterator:
            # Predecimos
            predictions = model(text.to(device))

            predictions = predictions.view(-1, predictions.shape[-1])
            tags = torch.reshape(tags, (-1,)).to(device)

            # Calculamos el Cross Entropy de las predicciones con respecto a las etiquetas reales
            loss = criterion(predictions, tags)

            # Calculamos las métricas
            precision, recall, f1 = calculate_metrics(predictions, tags)

            # Actualizamos el loss y las métricas
            epoch_loss += loss.item()
            epoch_precision += precision
            epoch_recall += recall
            epoch_f1 += f1

    return epoch_loss / len(iterator), epoch_precision / len(
        iterator), epoch_recall / len(iterator), epoch_f1 / len(iterator)

In [22]:
import time

def epoch_time(start_time, end_time):
    elapsed_time = end_time - start_time
    elapsed_mins = int(elapsed_time / 60)
    elapsed_secs = int(elapsed_time - (elapsed_mins * 60))
    return elapsed_mins, elapsed_secs

#### **Entrenamiento de la red**

En este cuadro de código ejecutaremos el entrenamiento de la red.
Para esto, primero definiremos el número de épocas y luego por cada época, ejecutaremos `train` y `evaluate`.

**Importante: Reiniciar los pesos del modelo**

Si ejecutas nuevamente esta celda, se seguira entrenando el mismo modelo una y otra vez.
Para reiniciar el modelo se debe ejecutar nuevamente la celda que contiene la función `init_weights`.

In [23]:
best_valid_loss = float('inf')

for epoch in range(n_epochs):

    start_time = time.time()

    # Recuerdo: dataloader_train y valid_iterator contienen el dataset dividido en batches.

    # Entrenar
    train_loss, train_precision, train_recall, train_f1 = train(
        model, dataloader_train, optimizer, criterion)

    # Evaluar (valid = validación)
    valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(
        model, dataloader_dev, criterion)

    end_time = time.time()

    epoch_mins, epoch_secs = epoch_time(start_time, end_time)

    # Si obtuvimos mejores resultados, guardamos este modelo en el almacenamiento (para poder cargarlo luego)
    # Si detienen el entrenamiento prematuramente, pueden cargar el modelo en el siguiente recuadro de código.
    if valid_loss < best_valid_loss:
        best_valid_loss = valid_loss
        torch.save(model.state_dict(), '{}.pt'.format(model_name))
    # Si ya no mejoramos el loss de validación, terminamos de entrenar.

    print(f'Epoch: {epoch+1:02} | Epoch Time: {epoch_mins}m {epoch_secs}s')
    print(
        f'\tTrain Loss: {train_loss:.3f} | Train f1: {train_f1:.2f} | Train precision: {train_precision:.2f} | Train recall: {train_recall:.2f}'
    )
    print(
        f'\t Val. Loss: {valid_loss:.3f} |  Val. f1: {valid_f1:.2f} |  Val. precision: {valid_precision:.2f} | Val. recall: {valid_recall:.2f}'
    )

/usr/local/lib/python3.10/dist-packages/seqeval/metrics/sequence_labeling.py:171: UserWarning: <PAD> seems not to be NE tag.
  warnings.warn('{} seems not to be NE tag.'.format(chunk))
/usr/local/lib/python3.10/dist-packages/seqeval/metrics/v1.py:57: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Epoch: 01 | Epoch Time: 0m 13s
	Train Loss: 0.836 | Train f1: 0.36 | Train precision: 0.54 | Train recall: 0.29
	 Val. Loss: 0.496 |  Val. f1: 0.66 |  Val. precision: 0.71 | Val. recall: 0.61
Epoch: 02 | Epoch Time: 0m 9s
	Train Loss: 0.397 | Train f1: 0.71 | Train precision: 0.76 | Train recall: 0.68
	 Val. Loss: 0.412 |  Val. f1: 0.71 |  Val. precision: 0.72 | Val. recall: 0.71
Epoch: 03 | Epoch Time: 0m 8s
	Train Loss: 0.258 | Train f1: 0.81 | Train precision: 0.82 | Train recall: 0.80
	 Val. Loss: 0.404 |  Val. f1: 0.72 |  Val. precision: 0.74 | Val. recall: 0.71
Epoch: 04 | Epoch Time: 0m 5s
	Train Loss: 0.191 | Train f1: 0.86 | Train precision: 0.87 | Train recall: 0.86
	 Val. Loss: 0.435 |  Val. f1: 0.72 |  Val. precision: 0.74 | Val. recall: 0.71
Epoch: 05 | Epoch Time: 0m 8s
	Train Loss: 0.145 | Train f1: 0.90 | Train precision: 0.90 | Train recall: 0.90
	 Val. Loss: 0.500 |  Val. f1: 0.72 |  Val. precision: 0.74 | Val. recall: 0.70
Epoch: 06 | Epoch Time: 0m 6s
	Train Loss: 0

**Importante**: Recuerden que el último modelo entrenado no es el mejor (probablemente esté *overfitteado*), si no el que guardamos con la menor loss del conjunto de validación. Este problema lo pueden solucionar con *early stopping*.
Para cargar el mejor modelo entrenado, ejecuten la siguiente celda.

In [24]:
# cargar el mejor modelo entrenado.
model.load_state_dict(torch.load('{}.pt'.format(model_name)))

<All keys matched successfully>

In [25]:
# Limpiar ram de cuda
torch.cuda.empty_cache()

#### **Evaluamos el set de validación con el modelo final**

Estos son los resultados de predecir el dataset de evaluación con el *mejor* modelo entrenado.

In [26]:
valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(
    model, dataloader_dev, criterion)

print(
    f'Val. Loss: {valid_loss:.3f} |  Val. f1: {valid_f1:.2f} | Val. precision: {valid_precision:.2f} | Val. recall: {valid_recall:.2f}'
)

Val. Loss: 0.404 |  Val. f1: 0.72 | Val. precision: 0.74 | Val. recall: 0.71


#### **Evaluamos el set de prueba con el modelo final**

Estos son los resultados de predecir el dataset de prueba con el *mejor* modelo entrenado.

In [27]:
prueba_loss, prueba_precision, prueba_recall, prueba_f1 = evaluate(
    model, dataloader_test, criterion)

print(
    f'Test. Loss: {prueba_loss:.3f} |  Test. f1: {prueba_f1:.2f} | Test. precision: {prueba_precision:.2f} | Test. recall: {prueba_recall:.2f}'
)

Test. Loss: 0.435 |  Test. f1: 0.71 | Test. precision: 0.73 | Test. recall: 0.69


## Experimentos

En esta sección deberan explicar, implementar y evaluar **TRES EXPERIMENTO DIFERENTES**. Puede revisar la [lista de sugerencias](#scrollTo=TfMlGTCgIq8o) al inicio del enunciado para sacar ideas de experimentos. Se espera que los experimentos sean relevantes y desafiantes.

Cada experimento debe contener las siguientes cuatro subsecciones:

- *Explicación del experimento*: debe motivar y explicar en que consiste su experimento. La explicación debe fundamentar porque es un experimento relevante y desafiantes, ademas de aclarar que estudiaran.

- *Implementacion del experimento*: debe implementar su experimento con codigo debidamente comentado, con tal de que sea legible.

- *Evaluar el experimento*: debe evaluar con las metricas propuestas para generar los resultados obtenidos con sus experimentos sobre el conjunto de prueba.

- *Analizar el experimento*: debe hacer un breve analisis de los resultados obtenidos.

### **Experimento 1**

#### Explicación

En este experimento se añadiran 2 nuevas capas al baseline para mejorar la tarea de reconocimiento de Entidades Nombradas en el baseline. Estas son:

- Capa de atención: Esta capa calcula pesos de atención, indicando que partes del input son más relevantes para la tarea actual.

- Capa de normalización: Una capa de normalización se utiliza en redes neuronales para normalizar las activaciones entre capas. Esto ayuda a mitigar los problemas de inestabilidad durante el entrenamiento y acelera la convergencia.

Además se agrego un drop out de 0.3.

En particular, nos interesa observar el efecto de la capa de atención en la tarea a resolver.

Utilizar capas de atención en modelos para resolver problemas NER es altamente beneficioso debido a su capacidad para capturar dependencias a largo plazo entre palabras dentro de una oración, facilitando así la identificación precisa de entidades nombradas. Estas capas permiten al modelo enfocarse selectivamente en partes específicas de la secuencia de entrada, lo que es crucial para gestionar contextos complejos y resolver ambigüedades sintácticas y semánticas. Además, en el panorama actual del aprendizaje profundo, las capas de atención han ganado prominencia significativa, especialmente con el desarrollo de modelos como los Transformers, que dependen en gran medida de mecanismos de atención para procesar eficientemente grandes cantidades de datos secuenciales y capturar relaciones complejas. Esta tecnología no solo mejora el rendimiento del modelo en términos de precisión, sino que también facilita la interpretación y la explicabilidad, permitiendo visualizar qué partes de la entrada son más relevantes para las decisiones del modelo, aspecto fundamental en aplicaciones sensibles como el análisis de sentimientos, diagnóstico médico y NER.





#### Implementación

In [68]:
class Attention(nn.Module):
    def __init__(self, hidden_dim):
        super().__init__()
        self.attn = nn.Linear(hidden_dim, hidden_dim)
        self.v = nn.Parameter(torch.rand(hidden_dim))

    def forward(self, hidden, encoder_outputs, mask=None):
        # hidden = [batch size, hid dim]
        # encoder_outputs = [src len, batch size, hid dim]

        src_len = encoder_outputs.shape[0]

        # repeat hidden state src_len times
        hidden = hidden.unsqueeze(1).repeat(1, src_len, 1)
        encoder_outputs = encoder_outputs.permute(1, 0, 2)

        # hidden = [batch size, src len, hid dim]
        # encoder_outputs = [batch size, src len, hid dim]

        energy = torch.tanh(self.attn(encoder_outputs))

        # energy = [batch size, src len, hid dim]

        energy = energy.permute(0, 2, 1)

        # energy = [batch size, hid dim, src len]

        v = self.v.repeat(encoder_outputs.size(0), 1).unsqueeze(1)

        # v = [batch size, 1, hid dim]

        attention = torch.bmm(v, energy).squeeze(1)

        # attention = [batch size, src len]

        if mask is not None:
            attention = attention.masked_fill(mask == 0, -1e10)

        return torch.softmax(attention, dim=1)

In [69]:
class NER_RNN_Attention(nn.Module):
    def __init__(self,
                 input_dim,
                 embedding_dim,
                 hidden_dim,
                 output_dim,
                 n_layers,
                 bidirectional,
                 dropout,
                 pad_idx):

        super().__init__()

        self.embedding = nn.Embedding(input_dim,
                                      embedding_dim,
                                      padding_idx=pad_idx)

        self.lstm = nn.LSTM(embedding_dim,
                            hidden_dim,
                            num_layers=n_layers,
                            bidirectional=bidirectional,
                            dropout=dropout if n_layers > 1 else 0)

        self.attention = Attention(hidden_dim * 2 if bidirectional else hidden_dim)

        self.layer_norm = nn.LayerNorm(hidden_dim * 2 if bidirectional else hidden_dim)


        self.fc = nn.Linear(hidden_dim * 2 if bidirectional else hidden_dim,
                            output_dim)

        self.dropout = nn.Dropout(dropout)

    def forward(self, text):

        # Mover text al mismo dispositivo que el embedding
        text = text.to(self.embedding.weight.device)

        embedded = self.embedding(text)
        embedded = self.dropout(embedded)

        outputs, (hidden, cell) = self.lstm(embedded)

        outputs = self.layer_norm(outputs)

        # aplicar mecanismo de atención
        attention_weights = self.attention(hidden[-1], outputs)
        attention_weights = attention_weights.unsqueeze(2)

        # aplicar pesos de atención
        attended_outputs = outputs.permute(1, 0, 2) * attention_weights
        attended_outputs = attended_outputs.permute(1, 0, 2)

        emissions = self.fc(attended_outputs)

        return emissions


In [70]:
### Escribir implenetación desde aquí
# Loss: Cross Entropy
TAG_PAD_IDX = NER_TAGS.vocab.get_stoi()['<PAD>']
baseline_criterion = nn.CrossEntropyLoss(ignore_index = TAG_PAD_IDX)


# tamaño del vocabulario. recuerden que la entrada son vectores bag of word(one-hot).
INPUT_DIM = len(TEXT.vocab)
EMBEDDING_DIM = 300  # dimensión de los embeddings.
HIDDEN_DIM = 256  # dimensión de la capas LSTM
OUTPUT_DIM = len(NER_TAGS.vocab)  # número de clases

N_LAYERS = 3  # número de capas.
DROPOUT = 0.3
BIDIRECTIONAL = False

# Creamos nuestro modelo.
exp_1 = NER_RNN_Attention(INPUT_DIM, EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM,
                         N_LAYERS, BIDIRECTIONAL, DROPOUT, PAD_IDX).to(device)

baseline_model_name = 'experimento_1'  # nombre que tendrá el modelo guardado...

baseline_n_epochs = 20


In [71]:
model_experimento_1 = exp_1
model_name = baseline_model_name
criterion = baseline_criterion
n_epochs = baseline_n_epochs

In [72]:
model_experimento_1.apply(init_weights)

NER_RNN_Attention(
  (embedding): Embedding(16496, 300, padding_idx=0)
  (lstm): LSTM(300, 256, num_layers=3, dropout=0.3)
  (attention): Attention(
    (attn): Linear(in_features=256, out_features=256, bias=True)
  )
  (layer_norm): LayerNorm((256,), eps=1e-05, elementwise_affine=True)
  (fc): Linear(in_features=256, out_features=12, bias=True)
  (dropout): Dropout(p=0.3, inplace=False)
)

In [73]:
optimizer = optim.Adam(model_experimento_1.parameters()) # SGD, AdamW

In [74]:
# Enviamos el modelo y la loss a cuda (en el caso en que esté disponible)
model_experimento_1 = model_experimento_1.to(device)
criterion = criterion.to(device)

In [75]:
#Entrenamiento

best_valid_loss = float('inf')

for epoch in range(n_epochs):

    start_time = time.time()

    # Recuerdo: dataloader_train y valid_iterator contienen el dataset dividido en batches.

    # Entrenar
    train_loss, train_precision, train_recall, train_f1 = train(
        model_experimento_1, dataloader_train, optimizer, criterion)

    # Evaluar (valid = validación)
    valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(
        model_experimento_1, dataloader_dev, criterion)

    end_time = time.time()

    epoch_mins, epoch_secs = epoch_time(start_time, end_time)

    # Si obtuvimos mejores resultados, guardamos este modelo en el almacenamiento (para poder cargarlo luego)
    # Si detienen el entrenamiento prematuramente, pueden cargar el modelo en el siguiente recuadro de código.
    if valid_loss < best_valid_loss:
        best_valid_loss = valid_loss
        torch.save(model_experimento_1.state_dict(), '{}.pt'.format(model_name))
    # Si ya no mejoramos el loss de validación, terminamos de entrenar.

    print(f'Epoch: {epoch+1:02} | Epoch Time: {epoch_mins}m {epoch_secs}s')
    print(
        f'\tTrain Loss: {train_loss:.3f} | Train f1: {train_f1:.2f} | Train precision: {train_precision:.2f} | Train recall: {train_recall:.2f}'
    )
    print(
        f'\t Val. Loss: {valid_loss:.3f} |  Val. f1: {valid_f1:.2f} |  Val. precision: {valid_precision:.2f} | Val. recall: {valid_recall:.2f}'
    )

Epoch: 01 | Epoch Time: 0m 6s
	Train Loss: 1.636 | Train f1: 0.01 | Train precision: 0.09 | Train recall: 0.01
	 Val. Loss: 1.052 |  Val. f1: 0.04 |  Val. precision: 0.30 | Val. recall: 0.02
Epoch: 02 | Epoch Time: 0m 7s
	Train Loss: 0.874 | Train f1: 0.30 | Train precision: 0.55 | Train recall: 0.22
	 Val. Loss: 0.751 |  Val. f1: 0.44 |  Val. precision: 0.57 | Val. recall: 0.37
Epoch: 03 | Epoch Time: 0m 6s
	Train Loss: 0.558 | Train f1: 0.60 | Train precision: 0.68 | Train recall: 0.53
	 Val. Loss: 0.563 |  Val. f1: 0.62 |  Val. precision: 0.66 | Val. recall: 0.60
Epoch: 04 | Epoch Time: 0m 7s
	Train Loss: 0.376 | Train f1: 0.74 | Train precision: 0.78 | Train recall: 0.71
	 Val. Loss: 0.516 |  Val. f1: 0.68 |  Val. precision: 0.70 | Val. recall: 0.67
Epoch: 05 | Epoch Time: 0m 6s
	Train Loss: 0.280 | Train f1: 0.81 | Train precision: 0.83 | Train recall: 0.80
	 Val. Loss: 0.502 |  Val. f1: 0.71 |  Val. precision: 0.74 | Val. recall: 0.68
Epoch: 06 | Epoch Time: 0m 6s
	Train Loss: 0.

In [76]:
# cargar el mejor modelo entrenado.
model_experimento_1.load_state_dict(torch.load('{}.pt'.format(model_name)))

<All keys matched successfully>

In [77]:
# Limpiar ram de cuda
torch.cuda.empty_cache()

#### Evaluación

In [78]:
prueba_loss, prueba_precision, prueba_recall, prueba_f1 = evaluate(
    model_experimento_1, dataloader_test, criterion)

print(
    f'Val. Loss: {prueba_loss:.3f} |  Val. f1: {prueba_f1:.2f} | Val. precision: {prueba_precision:.2f} | Val. recall: {prueba_recall:.2f}'
)

Val. Loss: 0.502 |  Val. f1: 0.70 | Val. precision: 0.74 | Val. recall: 0.67


#### Análisis

Observando los resultados se puede comprobar que el modelo del experimento 1 solo supera al baseline en la precisión.

Esto pudo ocurrir dado que las relaciones entre entidades nombradas y sus contextos no requieren de una atención detallada o si la estructura del problema es más dependiente de características locales o gramaticales que de relaciones globales dentro de las secuencias. Además, la implementación incorrecta de las capas de atención, como la elección inapropiada de hiperparámetros o la falta de suficiente cantidad de datos de entrenamiento, pudo llevar a un rendimiento inferior.

### **Experimento 2**

#### Explicación


En este experimento, se profundiza en la optimización de hiperparámetros para un modelo de Reconocimiento de Entidades Nombradas (NER) empleando una técnica de búsqueda en grilla. Exploraremos sistemáticamente la influencia de variar la dimensión de los embeddings, la dimensión de las capas ocultas LSTM, el número de capas LSTMy la configuración de bidireccionalidad en el rendimiento del modelo. Cada uno de estos parámetros tiene un impacto significativo en cómo el modelo aprende de los datos:

- Dimensión de Embedding (EMBEDDING_DIM): Afecta directamente la capacidad del modelo para representar y diferenciar características de los tokens en un espacio dimensional. Un espacio más amplio puede capturar más detalles, pero también puede provocar overfitting si es demasiado grande.

- Dimensión de las Capas LSTM (HIDDEN_DIM): Determina el tamaño de las unidades ocultas dentro de las capas LSTM. Unidades más grandes pueden modelar mejor la información y las dependencias temporales en los datos, pero incrementan el riesgo de overfitting y el costo computacional.

- Número de Capas LSTM (N_LAYERS): Incrementar el número de capas puede ayudar a aprender representaciones más complejas y abstractas de los datos. Sin embargo, también hace que el modelo sea más propenso al overfitting y más difícil de entrenar.

- Bidireccionalidad (BIDIRECTIONAL): Permite que las capas LSTM procesen la información en ambas direcciones (pasado y futuro) y es particularmente útil para comprender el contexto en problemas de secuencia como NER. Activar esta característica puede mejorar significativamente la precisión del modelo a costa de duplicar los recursos computacionales necesarios.

El objetivo del experimento es identificar la combinación óptima de estos hiperparámetros que maximice la eficacia del modelo, medido a través de métricas como el F1-score, precisión y recall. Esta búsqueda ayudará a mejorar la capacidad del modelo para procesar y etiquetar correctamente entidades en textos clínicos, proporcionando una herramienta más robusta y precisa para la extracción de información.

Notar que se decidió experimentar en una primera instancia, no usar dropout y adelantando los resultados se dejó así ya que no fue necesario porque se variando el resto de hiperparámetros se logró mejorar las 3 métricas objetivo F1-score, Precision y Recall. Además por una cuestión de tiempo solo se experimentó con la birideccionalidad activada ya que la hipótesis teórica dicta que debía mejorar las métricas, y commo ya dijimos se encontró una combinación de hiperparámetros con la birideccionalidad activada que mejoraba las 3 métricas objetivas no fue necesario probar con la bidireccionalidad desactivada.

#### Implementación

In [ ]:
from sklearn.model_selection import ParameterGrid
import copy
'''
EMBEDDING_DIM_EXP = 400  # Nueva dimensión de los embeddings.
HIDDEN_DIM_EXP = 512     # Nueva dimensión de las capas LSTM.
N_LAYERS_EXP = 4         # Nuevo número de capas LSTM.
BIDIRECTIONAL_EXP = True # Habilitar la bidireccionalidad.
'''
# Definimos un diccionario con los rangos de los hiperparámetros que queremos probar
param_grid = {
    'embedding_dim': [200, 300, 400],
    'hidden_dim': [256, 512],
    'n_layers': [2, 3, 4],
    'dropout': [0],
    'bidirectional': [True]
}

# Creamos la cuadrícula de parámetros
grid = ParameterGrid(param_grid)

# Mejor modelo inicial
best_model = None
best_f1 = 0
best_params = {}

# Evaluamos cada combinación de hiperparámetros
for params in grid:
    print(f"Probando configuración: {params}")

    # Establece la dimensión global de embedding para esta configuración
    global EMBEDDING_DIM
    EMBEDDING_DIM = params['embedding_dim']

    model = NER_RNN(INPUT_DIM, EMBEDDING_DIM, params['hidden_dim'], OUTPUT_DIM,
                    params['n_layers'], params['bidirectional'], params['dropout'], PAD_IDX)

    model.apply(init_weights)
    model = model.to(device)
    optimizer = optim.Adam(model.parameters())

    for epoch in range(n_epochs):
        train_loss, train_precision, train_recall, train_f1 = train(
            model, dataloader_train, optimizer, criterion)
        valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(
            model, dataloader_dev, criterion)

        print(f'Val. Loss: {valid_loss:.3f} | Val. F1: {valid_f1:.2f}')

    if valid_f1 > best_f1:
        best_f1 = valid_f1
        best_model = copy.deepcopy(model)
        best_params = params
        torch.save(best_model.state_dict(), 'best_model_exp2.pt')

print(f"Mejor F1: {best_f1:.2f} con configuración: {best_params}")


Probando configuración: {'bidirectional': True, 'dropout': 0, 'embedding_dim': 200, 'hidden_dim': 256, 'n_layers': 2}
Val. Loss: 0.397 | Val. F1: 0.72
Val. Loss: 0.349 | Val. F1: 0.75
Val. Loss: 0.370 | Val. F1: 0.76
Val. Loss: 0.480 | Val. F1: 0.75
Val. Loss: 0.481 | Val. F1: 0.77
Val. Loss: 0.523 | Val. F1: 0.76


KeyboardInterrupt: 

In [ ]:
import json
# Convertir el diccionario de los mejores parámetros a un string JSON
params_json = json.dumps(best_params, indent=4)

# Guardar el string JSON en un archivo
with open('best_params_exp2.json', 'w') as json_file:
    json_file.write(params_json)

print("Los mejores parámetros han sido guardados a 'best_params_exp2.json'")

Los mejores parámetros han sido guardados a 'best_params.json'


In [ ]:
print(f'Hiperparámetros del mejor modelo encontrado: {best_params}')

#### Evaluación

In [ ]:
# Cargamos el mejor modelo entrenado durante la búsqueda en cuadrícula
best_model.load_state_dict(torch.load('best_model_exp2.pt'))

# Enviamos el modelo a CUDA si está disponible
best_model = best_model.to(device)

# Evaluamos el modelo en el conjunto de prueba
test_loss, test_precision, test_recall, test_f1 = evaluate(
    best_model, dataloader_test, criterion)

print(f'Test Loss: {test_loss:.3f}')
print(f'Test F1-score: {test_f1:.2f}')
print(f'Test Precision: {test_precision:.2f}')
print(f'Test Recall: {test_recall:.2f}')


Final Validation Loss: 0.663
Final Validation F1-score: 0.77
Final Validation Precision: 0.79
Final Validation Recall: 0.75
Test Loss: 0.692
Test F1-score: 0.76
Test Precision: 0.79
Test Recall: 0.74


#### Análisis

Los resultados del experimento con la búsqueda en grilla de hiperparámetros demuestran una mejora significativa en el rendimiento general del modelo de Reconocimiento de Entidades Nombradas (NER) comparado con el modelo baseline. La configuración óptima identificada a través del experimento incluyó una arquitectura bidireccional con una dimensión de embedding reducida a 200, la misma dimensión de las capas LSTM de 256, y un menor número de capas LSTM, reduciendo de tres a dos.

Comparación de Rendimiento:

El modelo baseline logró un F1-score de 0.71, precisión de 0.73 y recall de 0.70. Esto se obtuvo con una configuración no bidireccional, una dimensión de embedding de 300, y tres capas LSTM.
El modelo optimizado, en cambio, alcanzó un F1-score de 0.76, con una precisión de 0.79 y un recall de 0.74. Esto representa un aumento de 5 puntos porcentuales en F1-score, 6 puntos en precisión y 4 puntos en recall.

En cuanto al impacto de los Hiperparámetros:

- Bidireccionalidad: La inclusión de la bidireccionalidad en el modelo optimizado fue decisiva para mejorar su capacidad de contextualización de las entidades en la secuencia de texto. Al procesar la información en ambas direcciones, el modelo pudo captar mejor las dependencias a largo plazo, lo que es especialmente útil en la tarea de NER donde el contexto puede influir significativamente en la clasificación de las entidades.

- Dimensión de Embedding: Reducir la dimensión de los embeddings de 300 a 200, contrario a la intuición inicial, no deterioró el rendimiento. Esta reducción puede haber ayudado a evitar el overfitting, permitiendo que el modelo generalizara mejor con menos parámetros a ajustar. Esto indica que una representación más compacta fue suficiente para captar las características relevantes del texto.

-Número de Capas LSTM: Reducir el número de capas de tres a dos también contribuyó a la mejora del rendimiento. En algunos casos, modelos más simples con menos capas pueden ser más efectivos, especialmente cuando el tamaño del dataset, como en este caso, no justifica una mayor complejidad. Esto puede ayudar a reducir el riesgo de overfitting y mejorar la velocidad de entrenamiento.

El experimento subraya la importancia de la configuración adecuada de los hiperparámetros y cómo ajustes aparentemente contra intuitivos pueden resultar en mejoras sustanciales. A pesar de reducir la dimensionalidad de los embeddings y el número de capas LSTM, la adición de la bidireccionalidad permitió una mejora notable en todas las métricas de evaluación. Este análisis sugiere que para futuros trabajos, explorar configuraciones que equilibren bien la complejidad y la capacidad de generalización del modelo puede ofrecer resultados aún mejores. Además, experimentar con diferentes tasas de dropout y otras técnicas de regularización podría proporcionar más perspectivas sobre cómo optimizar el rendimiento del modelo de NER en contextos clínicos y más allá.

### **Experimento 3**

#### Explicación

Este experimento propone reemplazar la arquitectura LSTM utilizada en el modelo baseline por una arquitectura GRU (Gated Recurrent Unit). Las GRU, siendo más simples estructuralmente que las LSTM, pueden ofrecer un rendimiento similar o incluso superior en ciertas tareas de procesamiento de lenguaje natural debido a su eficiencia en el manejo de secuencias de datos y menor complejidad de parámetros. Además de cambiar la arquitectura de la red, implementaremos técnicas de optimización de hiperparámetros mediante una grilla de búsqueda para determinar la mejor configuración posible en términos de dimensiones ocultas, número de capas, bidireccionalidad y dropout. También aumentaremos el número de épocas para el entrenamiento pero integraremos la técnica de early stopping para evitar el sobreajuste, asegurando así que el modelo generalice mejor en datos no vistos. Este enfoque se espera que supere al modelo baseline en términos de precisión, recall y puntuación F1. Este experimento es relevante ya que busca explorar cómo las diferencias estructurales entre LSTM y GRU afectan el rendimiento en una tarea compleja de NER, proporcionando insights sobre la configuración óptima de hiperparámetros en modelos de redes neuronales recurrentes.

#### Implementación

In [ ]:
import torch.nn as nn
import torch.optim as optim
from torch.optim.lr_scheduler import StepLR
from sklearn.model_selection import ParameterGrid
import copy

# Definir la red usando GRU
class NER_GRU(nn.Module):
    def __init__(self, input_dim, embedding_dim, hidden_dim, output_dim, n_layers, bidirectional, dropout, pad_idx):
        super().__init__()

        # Capa de embedding
        self.embedding = nn.Embedding(input_dim, embedding_dim, padding_idx=pad_idx)

        # Capa GRU
        self.gru = nn.GRU(embedding_dim, hidden_dim, num_layers=n_layers, bidirectional=bidirectional, dropout=dropout if n_layers > 1 else 0)

        # Capa de salida
        self.fc = nn.Linear(hidden_dim * 2 if bidirectional else hidden_dim, output_dim)

        # Dropout
        self.dropout = nn.Dropout(dropout)

    def forward(self, text):
        # text = [sent len, batch size]
        embedded = self.embedding(text)
        # embedded = [sent len, batch size, emb dim]
        outputs, hidden = self.gru(embedded)
        # outputs = [sent len, batch size, hid dim * num directions]
        predictions = self.fc(outputs)
        # predictions = [sent len, batch size, output dim]
        return predictions

# Parámetros para la búsqueda en grilla
param_grid = {
    'hidden_dim': [128, 256],
    'n_layers': [1, 2],
    'bidirectional': [True, False],
    'dropout': [0.5, 0.2],
    'learning_rate': [0.01, 0.001],
    'num_epochs': [10],
    'batch_size': [22]
}

# Función de early stopping
class EarlyStopping:
    def __init__(self, patience=5, min_delta=0):
        self.patience = patience
        self.min_delta = min_delta
        self.counter = 0
        self.best_loss = None
        self.early_stop = False

    def __call__(self, val_loss):
        if self.best_loss is None:
            self.best_loss = val_loss
        elif val_loss > self.best_loss - self.min_delta:
            self.counter += 1
            if self.counter >= self.patience:
                self.early_stop = True
        else:
            self.best_loss = val_loss
            self.counter = 0

best_model = None
best_f1 = 0
results = []

# Bucle para la búsqueda en grilla
for params in ParameterGrid(param_grid):
    print(params)
    model = NER_GRU(INPUT_DIM, EMBEDDING_DIM, params['hidden_dim'], OUTPUT_DIM, params['n_layers'], params['bidirectional'], params['dropout'], PAD_IDX)
    model.apply(init_weights)
    model = model.to(device)
    optimizer = optim.Adam(model.parameters(), lr=params['learning_rate'])
    criterion = nn.CrossEntropyLoss(ignore_index=TAG_PAD_IDX)
    criterion = criterion.to(device)
    early_stopping = EarlyStopping(patience=5)

    for epoch in range(params['num_epochs']):
        train_loss, train_precision, train_recall, train_f1 = train(model, dataloader_train, optimizer, criterion)
        valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(model, dataloader_dev, criterion)
        if early_stopping(valid_loss):
            break

    if valid_f1 > best_f1:
        best_f1 = valid_f1
        best_model = copy.deepcopy(model)
        best_params = params

    results.append({
        'params': params,
        'f1': valid_f1,
        'precision': valid_precision,
        'recall': valid_recall
    })

# Seleccionamos el mejor modelo
model = best_model

# Guardar el modelo
torch.save(model.state_dict(), 'best_gru_model.pt')
print("Mejor modelo guardado con F1:", best_f1)


{'batch_size': 22, 'bidirectional': True, 'dropout': 0.5, 'hidden_dim': 128, 'learning_rate': 0.01, 'n_layers': 1, 'num_epochs': 10}
{'batch_size': 22, 'bidirectional': True, 'dropout': 0.5, 'hidden_dim': 128, 'learning_rate': 0.01, 'n_layers': 2, 'num_epochs': 10}
{'batch_size': 22, 'bidirectional': True, 'dropout': 0.5, 'hidden_dim': 128, 'learning_rate': 0.001, 'n_layers': 1, 'num_epochs': 10}
{'batch_size': 22, 'bidirectional': True, 'dropout': 0.5, 'hidden_dim': 128, 'learning_rate': 0.001, 'n_layers': 2, 'num_epochs': 10}
{'batch_size': 22, 'bidirectional': True, 'dropout': 0.5, 'hidden_dim': 256, 'learning_rate': 0.01, 'n_layers': 1, 'num_epochs': 10}
{'batch_size': 22, 'bidirectional': True, 'dropout': 0.5, 'hidden_dim': 256, 'learning_rate': 0.01, 'n_layers': 2, 'num_epochs': 10}
{'batch_size': 22, 'bidirectional': True, 'dropout': 0.5, 'hidden_dim': 256, 'learning_rate': 0.001, 'n_layers': 1, 'num_epochs': 10}
{'batch_size': 22, 'bidirectional': True, 'dropout': 0.5, 'hidden_

In [ ]:
import json
# Nombre del archivo donde se guardarán los hiperparámetros
filename = 'best_gru_model_hyperparameters.json'

# Guardar los hiperparámetros en un archivo JSON
with open(filename, 'w') as file:
    json.dump(best_params, file, indent=4)

print(f'Hiperparámetros guardados en {filename}')

Hiperparámetros guardados en best_gru_model_hyperparameters.json


#### Evaluación

In [ ]:
# cargar el mejor modelo entrenado.
model.load_state_dict(torch.load('best_gru_model.pt'))

# Evaluamos el set de prueba con el modelo final
test_loss, test_precision, test_recall, test_f1 = evaluate(model, dataloader_test, criterion)

print(
    f'Test Loss: {test_loss:.3f} | Test F1: {test_f1:.2f} | Test Precision: {test_precision:.2f} | Test Recall: {test_recall:.2f}'
)


/home/ignacio/miniconda3/envs/nlp_t3/lib/python3.10/site-packages/torch/nn/modules/rnn.py:1133: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters(). (Triggered internally at ../aten/src/ATen/native/cudnn/RNN.cpp:1424.)
  result = _VF.gru(input, hx, self._flat_weights, self.bias, self.num_layers,


Test Loss: 0.699 | Test F1: 0.76 | Test Precision: 0.79 | Test Recall: 0.73


#### Análisis

El experimento con el modelo GRU ha demostrado ser bastante prometedor, superando al baseline con LSTM en todas las métricas de evaluación. La mejora en el rendimiento puede atribuirse a varios factores clave, entre los cuales la configuración de hiperparámetros y el cambio de arquitectura son los más significativos.

- Comparación de Arquitecturas y Configuraciones

    El baseline utilizando LSTM tenía configuraciones de hidden_dim=256, n_layers=3, y dropout=0, con un enfoque unidireccional. Por otro lado, el modelo GRU que mostró los mejores resultados utilizó una configuración de hidden_dim=128, n_layers=2, y dropout=0.2, y crucialmente, fue bidireccional.

    - Dimensiones y Capas Ocultas: Reducir la dimensión oculta y el número de capas en el modelo GRU, en comparación con el LSTM, podría haber contribuido a evitar el sobreajuste y mejorar la generalización, especialmente dado el mayor número de parámetros que requiere una configuración bidireccional.

    - Bidireccionalidad: La inclusión de la bidireccionalidad en el GRU permitió que el modelo capturara contextos tanto pasados como futuros de la secuencia, lo cual es una ventaja significativa para tareas de etiquetado secuencial como NER, donde la información contextual es crucial.

    - Dropout: La adición de un dropout de 0.2 ayudó a mitigar el sobreajuste durante el entrenamiento, permitiendo que el modelo generalizara mejor a nuevos datos. Esto contrasta con el baseline que no utilizaba dropout, lo que potencialmente podría llevar a un modelo que memoriza el conjunto de entrenamiento pero no generaliza bien.

- Resultados de Evaluación

    - LSTM (Baseline): Logró una precisión de 0.73, un recall de 0.70, y un F1-score de 0.71 con una pérdida de 0.421.
    - GRU (Optimizado): Mejoró significativamente, alcanzando una precisión de 0.79, un recall de 0.73, y un F1-score de 0.76, aunque con una pérdida más alta de 0.699.


La elección de GRU en este contexto particular ha demostrado ser bastante bueno, posiblemente debido a la simplicidad inherente de GRU, que tiene menos parámetros que administrar y es más fácil de optimizar con los recursos y el tamaño de datos disponibles.


- LSTM vs GRU

En comparación con el Experimento 2 donde se optimizó una LSTM, GRU tuvo los mismos resultados en las métricas objetivo a excepción del recall donde LSTM tuvo uno mas de recall con 0.74 vs 0.73, sin embargo, estos experimentos son insuficientes para decantar por uno o por otro, debido a que se entrenó por una cantidad muy limitada  de épocas y una cantidad limitada de opciones de valores para los hiperparámetros, sobre todo para la LSTM donde no se probó el dropout, sin embargo, al tener resultados tan similares, se puede declarar como ganador provisorio al GRU al ser un modelo mucho más simple y rápido de entrenar.